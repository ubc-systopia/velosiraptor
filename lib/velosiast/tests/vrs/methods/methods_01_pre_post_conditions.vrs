
// a requires caluse
fn foo() -> bool
    requires true
{
    true
}

// a ensures clause
fn foo() -> bool
    ensures true
{
    true
}

// both, ensures and requires
fn foo() -> bool
    requires true
    ensures true
{
    true
}

// method returning an int
fn bar() -> int {
    3
}

// referencing the parameter
fn baz(v: bool) -> bool
    requires v == true
{
    !v
}

// method returning an int
fn bar(v: bool, u: bool) -> bool
    requires v != u
{
    v && u
}

// more complex pre-post conditions. ensures will be ignored.
fn bar(val: int) -> int
    requires true;
    requires val > 0 && true;
    ensures bar(val) == val + 1;
    ensures true;
{
    1 + val
}
