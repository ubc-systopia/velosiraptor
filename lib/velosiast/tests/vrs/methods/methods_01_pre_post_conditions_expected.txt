  fn foo() -> bool
    requires true;
  {
    true
  }

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

  fn foo() -> bool
    requires true;
  {
    true
  }

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

  fn foo() -> bool
    requires true;
  {
    true
  }

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

  fn bar() -> int
    requires true;
  {
    0x3
  }

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

  fn baz(v: bool) -> bool
    requires v == true;
  {
    !(v)
  }

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

  fn bar(v: bool, u: bool) -> bool
    requires v != u;
  {
    v && u
  }

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

  fn bar(val: int) -> int
    requires true;
    requires 0x0 < val;
  {
    val + 0x1
  }

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

